File: jaeger_ai/interfaces/swift/Sources/JaegerOS/Bridge/Protocol.swift
//
//  Protocol.swift
//  JaegerOS / Bridge
//
//  Shared NDJSON value types + line framer for the stdio bridge. The bridge
//  speaks the JROS client protocol (see jaeger_os/protocol.py +
//  dev/docs/JROS_CLIENT_PROTOCOL.md); ``BridgeProcess`` parses the typed
//  frames itself. This file holds the small reusable pieces:
//
//    - ``Event``       — an inline activity event the AgentBridge fans out
//                        to ``ChatViewModel`` (thinking / tool chips).
//    - ``AnyDecodable`` — type-erased JSON value (Event payloads).
//    - ``FrameStream``  — splits a byte stream on '\n' into whole frames.
//
//  (The old socket-RPC Request/Response/NDJSON codec was removed with the
//  daemon — BridgeProcess reads NDJSON frames directly.)
//

import Foundation

/// An inline activity event surfaced during a turn (e.g. thinking,
/// tool.start / tool.complete). ``payload`` carries event-specific fields.
///
/// ``@unchecked Sendable`` is honest: the wrapped values come from
/// ``JSONSerialization`` (immutable Foundation types) and the struct is a
/// ``let``-only carrier.
struct Event: @unchecked Sendable {
    let name: String
    let payload: [String: AnyDecodable]
}

/// Type-erased JSON value — pull a typed value out with ``get(_:)``.
///
/// ``@unchecked Sendable`` is fine for the same reason as ``Event``: the
/// wrapped value is always a JSONSerialization-produced immutable type.
struct AnyDecodable: @unchecked Sendable {
    let value: Any

    init(_ value: Any) { self.value = value }

    /// Pull out the value if it is of the expected type, else nil.
    func get<T>(_ type: T.Type) -> T? { value as? T }
}

/// Stateful framer: append inbound bytes, get back whole NDJSON frames.
/// Partial frames buffer until a later ``feed`` completes them (pipe reads
/// aren't aligned to message boundaries).
final class FrameStream {
    private var buffer = Data()

    /// Append bytes; return any complete frames now parseable (no newline).
    func feed(_ chunk: Data) -> [Data] {
        buffer.append(chunk)
        var frames: [Data] = []
        while let nl = buffer.firstIndex(of: 0x0A) {
            let frame = buffer.subdata(in: buffer.startIndex..<nl)
            buffer.removeSubrange(buffer.startIndex...nl)
            if !frame.isEmpty { frames.append(frame) }
        }
        return frames
    }
}

// MARK: - Protocol v1 typed frames

/// The protocol version this shell speaks. Compared against the ``proto``
/// field in ``ready`` — a mismatch is surfaced, never silently degraded.
enum ProtocolV1 {
    static let version = "1"
}

/// Every agent→client frame, decoded strictly by its ``type`` discriminator.
/// THE typed mirror of ``jaeger_os/contract/protocol.py`` — shapes are
/// pinned by ProtocolFixtureTests against ``protocol_v1_fixtures.json``,
/// the same fixture file the Python side asserts its builders against.
/// An unknown ``type`` decodes to nil (forward-compatible: new frames from
/// a newer core are skipped, not fatal).
enum ProtocolFrame {
    case ready(BridgeReady)
    case agentState(AgentLifecycle)
    case state(busy: Bool)
    /// ``detail`` is a v1 ADDITIVE optional — short human context for the
    /// activity chip (today: which skill loaded, e.g. "view scheduling").
    case tool(name: String, phase: String, elapsed: Double, detail: String?)
    /// ``telemetry`` fields are v1 ADDITIVE optionals — a core that
    /// doesn't send them (or an older fixture) decodes to nils.
    case reply(text: String, error: String?,
               elapsedS: Double?, ctxUsed: Int?, ctxMax: Int?)
    case result(id: String, ok: Bool, error: String?, data: Data?)
    case request(BridgeRequest)
    /// ``suggestedName`` is a v1 ADDITIVE optional (nil when the core
    /// omits it, or on an older core) — the operator's CLI-pinned agent
    /// name, sent alongside ``kind="no_instance"`` so onboarding can
    /// default the identity step to it.
    case fatal(error: String, kind: String, suggestedName: String?)
    case bye

    static func decode(_ frame: Data) -> ProtocolFrame? {
        guard let obj = (try? JSONSerialization.jsonObject(with: frame))
                as? [String: Any],
              let type = obj["type"] as? String
        else { return nil }
        switch type {
        case "ready":
            return .ready(BridgeReady(
                instance: obj["instance"] as? String ?? "",
                model: obj["model"] as? String,
                character: obj["character"] as? String,
                icon: obj["icon"] as? String,
                proto: obj["proto"] as? String ?? "0",
                capabilities: obj["capabilities"] as? [String] ?? [],
                agent: obj["agent"] as? String ?? "ready",
                agentName: obj["agent_name"] as? String))
        case "agent_state":
            switch obj["state"] as? String ?? "" {
            case "booting":
                return .agentState(.booting)
            case "ready":
                return .agentState(.ready(model: obj["model"] as? String,
                                          character: obj["character"] as? String,
                                          icon: obj["icon"] as? String,
                                          agentName: obj["agent_name"] as? String))
            case "failed":
                return .agentState(.failed(obj["error"] as? String ?? "agent failed"))
            default:
                return nil
            }
        case "state":
            return .state(busy: obj["busy"] as? Bool ?? false)
        case "tool":
            return .tool(name: obj["name"] as? String ?? "",
                         phase: obj["phase"] as? String ?? "start",
                         elapsed: (obj["elapsed_s"] as? Double) ?? 0,
                         detail: obj["detail"] as? String)
        case "reply":
            return .reply(text: obj["text"] as? String ?? "",
                          error: obj["error"] as? String,
                          elapsedS: (obj["elapsed_s"] as? NSNumber)?.doubleValue,
                          ctxUsed: (obj["ctx_used"] as? NSNumber)?.intValue,
                          ctxMax: (obj["ctx_max"] as? NSNumber)?.intValue)
        case "result":
            var payload: Data? = nil
            if let d = obj["data"], !(d is NSNull) {
                payload = try? JSONSerialization.data(withJSONObject: d,
                                                      options: [.fragmentsAllowed])
            }
            return .result(id: obj["id"] as? String ?? "",
                           ok: obj["ok"] as? Bool ?? true,
                           error: obj["error"] as? String, data: payload)
        case "request":
            return .request(BridgeRequest(
                id: obj["id"] as? String ?? "",
                kind: obj["kind"] as? String ?? "approval",
                prompt: obj["prompt"] as? String ?? "",
                options: obj["options"] as? [String] ?? []))
        case "fatal":
            return .fatal(error: obj["error"] as? String ?? "bridge failed",
                          kind: obj["kind"] as? String ?? "boot",
                          suggestedName: obj["suggested_name"] as? String)
        case "bye":
            return .bye
        default:
            return nil
        }
    }
}

File: jaeger_os/contract/protocol.py
"""JROS client protocol — the single wire contract for surfaces.

Lives in ``jaeger_os.contract`` (0.9 contract package) — frame shapes are
wire truth, not interface implementation. ``protocol_v1_fixtures.json``
lives beside this file; ``jaeger_os.interfaces.bridge`` builds these frames
and ``jaeger_os.interfaces.client.JrosClient`` parses them.

"Transports, not endpoints": one protocol, many transports. The Swift app
speaks it over stdio (``jaeger bridge``); the MCP server, a web backend, or
any third-party client speaks the *same* frames over the *same* SDK
(:class:`jaeger_os.interfaces.client.JrosClient`). A future WebSocket transport bridges
the in-process chassis bus to these identical frames — so a new surface is a
transport, not a re-implementation.

Frames are JSON objects, one per line (NDJSON):

  client → agent:
    {"op": "send",    "text": <str>, "session": <str?>}     # run one turn
    {"op": "respond", "id": <str>,   "answer": <str>}       # answer a request
    {"op": "quit"}                                          # graceful stop

  agent → client:
    {"type": "ready",   "instance": <str>, "model": <str?>}
    {"type": "state",   "busy": <bool>,    "session": <str?>}
    {"type": "tool",    "name": <str>, "phase": <start|done|error>,
                        "elapsed_s": <float>, "session": <str?>}
    {"type": "reply",   "text": <str>, "error": <str?>, "session": <str?>}
    {"type": "request", "id": <str>, "kind": <approval|clarify|secret>,
                        "prompt": <str>, "options": [<str>...], "session": <str?>}
    {"type": "fatal",   "error": <str>}                     # boot failed

This module is the ONE place these shapes live: the bridge builds them, the
client parses them, and the bus↔wire codec (``event_to_frame``) maps chassis
messages onto them. ``PROTOCOL_VERSION`` bumps on any breaking change.

``query``/``command`` are generic envelopes — ``what``/``cmd`` name the verb,
``args`` carries its payload, and adding one is additive (no version bump,
one branch in ``bridge.py``'s ``_query``/``_command``/``main``). The native
History surface (runway item 4, 0.8) added three:

  query   ``list_sessions`` {limit?}     -> [{id, title, preview,
                                              created_at, last_active,
                                              messages}, ...]
  query   ``load_session``  {id}         -> [{role, text, ts}, ...]  (also
                                             replays into the live agent —
                                             see ``main.resume_session_from_store``)
  command ``new_session``   {old_id?}    -> {id: <new session id>}  (evicts
                                             old_id when given)

In-app updates (0.8) added two more:

  query   ``check_update``  {}          -> {current, latest, available,
                                             notes_url}  (version_check.
                                             cached_update_status — cached
                                             ~6h under <instance>/run/, so a
                                             tray poll is cheap; never
                                             raises, degrades to
                                             available:false offline)
  command ``run_update``    {ref?}      -> {restart_required, returncode,
                                             output}  (shells out to
                                             `jaeger update`, the SAME
                                             machinery the CLI verb runs;
                                             refuses with an error while a
                                             turn is in flight)
"""

from __future__ import annotations

import json
from typing import Any

PROTOCOL_VERSION = "1"

# What this transport can do — sent in ``ready`` so a client can feature-gate
# instead of probing. A client seeing an unknown capability ignores it; a
# client MISSING one it needs degrades that feature, not the connection.
CAPABILITIES: tuple[str, ...] = (
    "query", "command", "chat", "sessions", "permissions", "agent_state",
)

# ── agent → client frame builders (used by the bridge / any transport) ──


def ready_frame(instance: str, model: str | None,
                character: str | None = None, icon: str | None = None,
                agent: str = "ready", agent_name: str | None = None) -> dict[str, Any]:
    # ``character`` = the active character's display name; ``icon`` = an absolute
    # path to its profile image. Both let the native client show the agent's face
    # + name in the tray/header, matching the PySide6 UI.
    #
    # v1 additions: ``proto`` + ``capabilities`` (so shell/core version skew
    # fails loudly instead of degrading silently) and ``agent`` — the agent
    # lifecycle state at handshake time. The bridge now emits ``ready`` the
    # moment the TRANSPORT is usable (queries/commands work immediately);
    # ``agent`` says whether the model is loaded ("ready") or still coming up
    # ("booting"), with ``agent_state`` frames streaming the transition.
    return {"type": "ready", "proto": PROTOCOL_VERSION,
            "capabilities": list(CAPABILITIES),
            "instance": instance, "model": model,
            "character": character, "icon": icon, "agent": agent,
            "agent_name": agent_name}


def agent_state_frame(state: str, model: str | None = None,
                      character: str | None = None, icon: str | None = None,
                      error: str | None = None,
                      agent_name: str | None = None) -> dict[str, Any]:
    """The agent lifecycle, decoupled from transport readiness:
    ``booting`` → ``ready`` (model/character attached) or ``failed``
    (``error`` says why; ``kind`` distinguishes a held instance lock).

    ``agent_name`` (v1 additive) is the AGENT's own name (identity.yaml — the
    unique robot the operator named); ``character`` is the persona it plays.
    Surfaces lead with ``agent_name`` so ``ready`` doesn't flash the character
    name before the ``identity`` query resolves. ``icon`` is the effective
    avatar (instance profile picture if set, else the character card)."""
    return {"type": "agent_state", "state": state, "model": model,
            "character": character, "icon": icon, "error": error,
            "agent_name": agent_name}


def state_frame(busy: bool, session: str = "") -> dict[str, Any]:
    return {"type": "state", "busy": busy, "session": session}


def result_frame(req_id: Any, data: Any = None, ok: bool = True,
                 error: str | None = None) -> dict[str, Any]:
    # Reply to a native client's {"op":"query"|"command", "id":…}. ``data`` holds
    # the query payload; ``ok``/``error`` report command success.
    return {"type": "result", "id": req_id, "ok": ok, "data": data, "error": error}


def tool_frame(name: str, phase: str, elapsed_s: float = 0.0,
               session: str = "", *, detail: str = "") -> dict[str, Any]:
    """``detail`` (v1 ADDITIVE, omitted when empty — clients must decode
    frames without it): short human context for the activity chip. Today
    only ``skill`` calls set it ("view scheduling"), so surfaces can show
    WHICH skill loaded instead of a bare tool name."""
    frame: dict[str, Any] = {"type": "tool", "name": name, "phase": phase,
                             "elapsed_s": float(elapsed_s), "session": session}
    if detail:
        frame["detail"] = detail
    return frame


def reply_frame(text: str, error: str | None = None,
                session: str = "", *,
                elapsed_s: float | None = None,
                ctx_used: int | None = None,
                ctx_max: int | None = None) -> dict[str, Any]:
    """One finished turn. v1 ADDITIVE telemetry (optional — the keys are
    OMITTED when unknown, and clients must decode frames without them):

      ``elapsed_s``  wall-clock seconds the turn took ("replied in 3s")
      ``ctx_used``   estimated prompt tokens the session occupies now
      ``ctx_max``    the loaded model's context window ("ctx 18.3K/32.8K")
    """
    frame: dict[str, Any] = {"type": "reply", "text": text, "error": error,
                             "session": session}
    if elapsed_s is not None:
        frame["elapsed_s"] = round(float(elapsed_s), 2)
    if ctx_used is not None:
        frame["ctx_used"] = int(ctx_used)
    if ctx_max is not None:
        frame["ctx_max"] = int(ctx_max)
    return frame


def request_frame(id: str, kind: str, prompt: str,
                   options: tuple[str, ...] | list[str] = (),
                   session: str = "") -> dict[str, Any]:
    return {"type": "request", "id": id, "kind": kind, "prompt": prompt,
            "options": list(options), "session": session}


def fatal_frame(error: str, kind: str = "boot",
                *, suggested_name: str | None = None) -> dict[str, Any]:
    """Unrecoverable failure — the bridge exits after this. ``kind``:
    ``boot`` (agent failed to start), ``locked`` (another process holds
    this instance's lock — the client should offer attach-or-pick-another,
    not a generic error), or ``no_instance`` (v1 additive: the resolved
    instance doesn't exist on disk yet — first-run. The transport STAYS
    alive for queries/commands so a native client can run onboarding and
    ``create_instance`` over the same pipe).

    ``suggested_name`` (v1 ADDITIVE, omitted when absent — clients must
    decode frames without it): only sent alongside ``kind="no_instance"``,
    the operator-pinned instance name (e.g. ``./jaeger agent create
    lilith``'s ``JAEGER_INSTANCE_NAME`` pin) so onboarding can default the
    identity step's name field to it instead of silently orphaning it
    behind whatever character the operator picks."""
    frame: dict[str, Any] = {"type": "fatal", "error": error, "kind": kind}
    if suggested_name:
        frame["suggested_name"] = suggested_name
    return frame


def queued_frame(session: str, position: int) -> dict[str, Any]:
    """v1 ADDITIVE (0.8.1 item 9): a ``send`` landed while a turn was already
    in flight. The bridge NEVER drops it — one worker thread drains a FIFO
    queue, so a send that arrives mid-turn just runs as the next normal
    turn (its own ``state``/``reply`` frames follow once it's up). This
    frame is purely a VISIBILITY signal so a client can render "queued"
    instead of assuming nothing happened. ``position`` counts THIS
    session's outstanding queued sends including this one (1 = next up)."""
    return {"type": "queued", "session": session, "position": int(position)}


def bye_frame(reason: str = "quit") -> dict[str, Any]:
    """Clean-shutdown marker: emitted right before the bridge exits on a
    ``quit`` op (or EOF), so the client can tell an ORDERLY exit from a
    crash even though the process may exit through ``os._exit`` (the ggml
    Metal teardown makes exit codes unreliable — see F1 in STATUS.md)."""
    return {"type": "bye", "reason": reason}


# ── client → agent op builders (used by the client SDK) ──


def send_op(text: str, session: str = "") -> dict[str, Any]:
    return {"op": "send", "text": text, "session": session}


def respond_op(id: str, answer: str) -> dict[str, Any]:
    return {"op": "respond", "id": id, "answer": answer}


def quit_op() -> dict[str, Any]:
    return {"op": "quit"}


# ── parsing ──


def parse(line: str) -> dict[str, Any] | None:
    """Parse one NDJSON line into a frame dict, or None if malformed / not a
    protocol frame (no ``type``/``op`` discriminator)."""
    line = line.strip()
    if not line:
        return None
    try:
        obj = json.loads(line)
    except (json.JSONDecodeError, ValueError):
        return None
    if not isinstance(obj, dict) or not ("type" in obj or "op" in obj):
        return None
    return obj


def encode(frame: dict[str, Any]) -> str:
    """Serialize a frame to one NDJSON line (newline included)."""
    return json.dumps(frame, ensure_ascii=False) + "\n"


# ── chassis-bus message → wire frame (for bus-attached transports) ──


def event_to_frame(msg: Any) -> dict[str, Any] | None:
    """Map a chassis ``/sense/*`` message to a wire frame, or None if the
    message isn't part of the client protocol. Lets a WebSocket/socket
    transport mirror the in-process bus to remote clients identically."""
    topic = getattr(msg, "topic", "")
    session = getattr(msg, "session", "") or ""
    if topic == "/sense/chat":
        return reply_frame(getattr(msg, "text", ""), None, session)
    if topic == "/sense/agent_state":
        return state_frame(getattr(msg, "state", "") == "thinking", session)
    if topic == "/sense/tool":
        return tool_frame(getattr(msg, "name", ""), getattr(msg, "phase", "start"),
                          getattr(msg, "elapsed_s", 0.0), session,
                          detail=getattr(msg, "detail", "") or "")
    if topic == "/sense/request":
        return request_frame(getattr(msg, "id", ""), getattr(msg, "kind", "approval"),
                             getattr(msg, "prompt", ""),
                             getattr(msg, "options", ()), session)
    return None
